File: ruby/lib/pay_kit/price.rb
# frozen_string_literal: true

require "bigdecimal"

require_relative "errors"

module PayKit
  # A single settlement preference: pay `amount` denominated in `coin`.
  # v1 always sets `amount` equal across a `Price`'s settlements; the
  # shape leaves room for per-coin overrides later (rule 5 in design).
  Settlement = Data.define(:coin, :amount) do
    def to_s
      "#{amount} #{coin}"
    end
  end

  # Denomination + ordered settlement preference list.
  #
  #   Price.new(denom: :USD, amount: "0.10", settlements: [Settlement(coin: :USDC, amount: "0.10")])
  #
  # Build via the `usd("0.10", :USDC, :USDT)` shorthand (see
  # PayKit::Helpers::Pricing). `settlements` is always non-empty; the
  # first entry is the preference.
  class Price
    attr_reader :denom, :amount, :settlements

    def initialize(denom:, amount:, settlements:)
      raise ConfigurationError, "denom must be a symbol like :USD" unless denom.is_a?(Symbol)
      raise ConfigurationError, "amount must be a non-empty string" unless amount.is_a?(String) && !amount.empty?
      raise ConfigurationError, "settlements must be a non-empty array" if !settlements.is_a?(Array) || settlements.empty?
      unless settlements.all? { |s| s.is_a?(Settlement) }
        raise ConfigurationError, "settlements must be Settlement objects"
      end

      @denom = denom
      @amount = amount
      @settlements = settlements.freeze
      freeze
    end

    # Build a Price denominated in `denom` from the variadic
    # `coins` list. Empty list means "use config defaults" and
    # is filled in later by the Pricing DSL.
    def self.build(denom:, amount:, coins: [])
      settlements = coins.flatten.compact.map { |coin| Settlement.new(coin: coin.to_sym, amount: amount) }
      new(denom: denom, amount: amount, settlements: settlements)
    end

    # The primary settlement coin (first preference). Used by
    # single-recipient flows where only the top choice matters.
    # Settlements is guaranteed non-empty by `Price.new`.
    def primary_coin
      settlements.first.coin
    end

    def to_s
      "#{denom} #{amount} (#{settlements.map(&:coin).join(", ")})"
    end

    # Numeric amount for fee math. BigDecimal-precise. Recomputed
    # per call so the frozen Price stays frozen.
    def to_d
      raise ConfigurationError, "invalid amount: #{amount.inspect}" unless /\A\d+(\.\d+)?\z/.match?(amount)

      BigDecimal(amount)
    end

    # Build a new Price with the same denom and a different amount.
    # Settlements list reuses the existing coin order.
    def with_amount(new_amount)
      Price.new(
        denom: denom,
        amount: new_amount.to_s,
        settlements: settlements.map { |s| Settlement.new(coin: s.coin, amount: new_amount.to_s) }
      )
    end
  end

  module Helpers
    # Mixed into the Pricing DSL and into controller helpers so
    # `usd("0.10")` works in both call sites. When no coins are
    # passed explicitly, falls back to `PayKit.config.stablecoins`.
    module Pricing
      def usd(amount, *coins)
        ::PayKit::Helpers::Pricing.build_price(:USD, amount, coins)
      end

      def eur(amount, *coins)
        ::PayKit::Helpers::Pricing.build_price(:EUR, amount, coins)
      end

      def gbp(amount, *coins)
        ::PayKit::Helpers::Pricing.build_price(:GBP, amount, coins)
      end

      def self.build_price(denom, amount, coins)
        resolved = coins.flatten.compact
        if resolved.empty?
          resolved = ::PayKit.config.stablecoins
          if resolved.empty?
            raise ::PayKit::ConfigurationError,
              "no stablecoins specified and PayKit.config.stablecoins is empty"
          end
        end
        ::PayKit::Price.build(denom: denom, amount: amount.to_s, coins: resolved)
      end
    end
  end
end

File: python/src/pay_kit/price.py
"""Denominated amount plus an ordered settlement-preference list.

``Price.usd("0.10")`` reads "ten cents USD, settle in whatever the config
prefers". ``Price.usd("0.10", Stablecoin.USDC)`` narrows settlement to USDC.
The variadic stablecoin order is preference; the resolver picks the first coin
it can settle.

Amounts are :class:`decimal.Decimal`. The factories reject ``float`` at the
signature level (accept ``str | int | Decimal`` only) so binary-float rounding
never touches money. Build via :meth:`Price.usd` / :meth:`Price.eur` /
:meth:`Price.gbp`, never the raw constructor.
"""

from __future__ import annotations

import re
from decimal import Decimal, InvalidOperation

import pydantic

from pay_kit._paycore.currency import Currency
from pay_kit._paycore.stablecoin import Stablecoin
from pay_kit.errors import ConfigurationError, MixedCurrenciesError

__all__ = ["Price", "Settlement"]

_AMOUNT_RE = re.compile(r"^\d+(\.\d+)?$")


def _to_decimal(amount: object) -> Decimal:
    """Coerce a money input to Decimal, rejecting float and bad formats.

    Accepts ``object`` (not just ``str | int | Decimal``) because the public
    factories forward untyped caller input and the field validator forwards a
    raw pydantic value; the isinstance ladder is the load-bearing runtime guard.
    """
    if isinstance(amount, bool):  # bool is an int subclass; reject explicitly.
        raise ConfigurationError("pay_kit: Price amount must be str | int | Decimal, not bool")
    if isinstance(amount, float):
        raise ConfigurationError("pay_kit: Price amount must be str | int | Decimal, not float")
    if isinstance(amount, Decimal):
        coerced = amount
    elif isinstance(amount, int):
        coerced = Decimal(amount)
    elif isinstance(amount, str):
        if not _AMOUNT_RE.match(amount):
            raise ConfigurationError(f"pay_kit: invalid Price amount: {amount!r}")
        try:
            coerced = Decimal(amount)
        except InvalidOperation as exc:
            raise ConfigurationError(f"pay_kit: invalid Price amount: {amount!r}") from exc
    else:
        raise ConfigurationError("pay_kit: Price amount must be str | int | Decimal")
    # The str path rejects negatives via _AMOUNT_RE, but the int and Decimal
    # paths reach here unguarded; validate the sign uniformly so usd(-1) and
    # usd(Decimal("-0.01")) raise instead of building an invalid Price. Zero is
    # allowed (a free gate).
    if coerced < 0:
        raise ConfigurationError(f"pay_kit: Price amount must not be negative: {amount!r}")
    return coerced


class Settlement(pydantic.BaseModel):
    """A single settlement preference: pay ``amount`` denominated in ``coin``."""

    model_config = pydantic.ConfigDict(frozen=True, extra="forbid")

    coin: Stablecoin
    amount: str

    def __str__(self) -> str:
        return f"{self.amount} {self.coin.value}"


class Price(pydantic.BaseModel):
    """Currency-denominated amount with an ordered settlement-coin preference."""

    model_config = pydantic.ConfigDict(frozen=True, extra="forbid")

    amount: Decimal
    currency: Currency
    settlements: tuple[Stablecoin, ...] = ()

    @pydantic.field_validator("amount", mode="before")
    @classmethod
    def _coerce_amount(cls, value: object) -> Decimal:
        return _to_decimal(value)

    @classmethod
    def usd(cls, amount: str | int | Decimal, *settlements: Stablecoin) -> Price:
        """Build a USD-denominated price."""
        return cls(amount=_to_decimal(amount), currency=Currency.USD, settlements=settlements)

    @classmethod
    def eur(cls, amount: str | int | Decimal, *settlements: Stablecoin) -> Price:
        """Build a EUR-denominated price."""
        return cls(amount=_to_decimal(amount), currency=Currency.EUR, settlements=settlements)

    @classmethod
    def gbp(cls, amount: str | int | Decimal, *settlements: Stablecoin) -> Price:
        """Build a GBP-denominated price."""
        return cls(amount=_to_decimal(amount), currency=Currency.GBP, settlements=settlements)

    def with_amount(self, amount: str | int | Decimal) -> Price:
        """Return a copy with a new amount, same currency and settlements."""
        return Price(amount=_to_decimal(amount), currency=self.currency, settlements=self.settlements)

    def plus(self, other: Price) -> Price:
        """Sum two same-currency prices; raise on a currency mismatch."""
        if self.currency != other.currency:
            raise MixedCurrenciesError(
                f"pay_kit: cannot sum prices of different currencies ({self.currency.value} vs {other.currency.value})"
            )
        return Price(
            amount=self.amount + other.amount,
            currency=self.currency,
            settlements=self.settlements,
        )

    def amount_string(self) -> str:
        """The wire-form decimal string, preserving trailing zeros."""
        return format(self.amount, "f")

    def primary_coin(self) -> Stablecoin | None:
        """The most-preferred settlement coin, or None to defer to config."""
        return self.settlements[0] if self.settlements else None
